### DISCUSSION TITLE
TLS 1.2 Restriction Impacting Development Environments in Projects Relying on Newer Node.js Versions

### DISCUSSION THREAD

**Discussion by Neizan93** [openssl/openssl#22537] [Q&A]
### Discussed in https://github.com/openssl/openssl/discussions/22520

<div type='discussions-op-text'>

<sup>Originally posted by **Neizan93** October 26, 2023</sup>
I am working on a project utilizing Angular 15 and Node 18 in a local development environment. A critical part of our development workflow involves the use of proxy configuration to point to different backends from the local environment. However, we encountered an issue where the proxy configuration fails to function when targeting HTTPS backends using TLS 1.2.

On investigating, it was identified that the root cause lies in the change in OpenSSL version used by Node.js from version 17 onwards. Node 18, which utilizes OpenSSL 3, restricts connections to remote endpoints using TLS versions lower than 1.3 as it's deemed insecure. This restriction, while understandable from a security standpoint, significantly disrupts our development workflow as our backends are currently configured with TLS 1.2, with no immediate plans to upgrade to TLS 1.3.

OpenSSL Version
`OpenSSL 3.0.0`

**Issue Reproduction**
Set up a local development environment with Angular 15 and Node 18.
Configure the proxy settings in Angular CLI to point to a backend using HTTPS with TLS 1.2.
Attempt to access the backend via the proxy from the local development environment.
Observe the failure in proxy configuration, preventing access to the backend due to TLS 1.2 restriction enforced by OpenSSL 3.
Expected Behavior
A mechanism or configuration option within OpenSSL that allows for temporary override of the TLS version restriction, especially in a controlled, local development environment, would be highly beneficial. This feature would enable developers to work with backends that are yet to upgrade to TLS 1.3, ensuring a smoother development workflow.

**Additional Context**
The restriction on TLS versions impacts not just our project but potentially many other development environments relying on frameworks and tools that utilize newer versions of Node.js and, by extension, OpenSSL 3.</div>
```[tasklist]
### Tasks
```

**Comment by nhorman** [comment]
As was noted in #22520, openssl doesn't restrict the use of tls 1.2, its all configurable.  I presume your backend is the system running node.js?  If so, node.js environment has a setting in the tls api (that is likely set by default to tls1.3):

https://nodejs.org/docs/latest-v17.x/api/tls.html#tlsdefault_min_version

Setting it to tls1.2 should allow for connections from tls1.2 clients
